runners/local-artifact-service-java/src/main/java/org/apache/beam/artifact/local/LocalFileSystemArtifactStagerService.java

```diff
@@ -0,0 +1,270 @@
+    if (stagingBase.exists() && stagingBase.canWrite()) {
+      artifactsBase = new File(stagingBase, "artifacts");
+      artifactsBase.mkdir();
+    } else {
+      throw new IllegalStateException(
+          String.format("Cannot write to base directory %s", stagingBase));
+    }
+    checkState(
+        stagingBase.exists() && stagingBase.canWrite(),
+        "Could not create staging directory structure at root %s",
+        stagingBase);
+  }
+
+  @Override
+  public StreamObserver<PutArtifactRequest> putArtifact(
+      final StreamObserver<PutArtifactResponse> responseObserver) {
+    return new CreateAndWriteFileObserver(responseObserver);
+  }
+
+  @Override
+  public void commitManifest(
+      CommitManifestRequest request, StreamObserver<CommitManifestResponse> responseObserver) {
+    try {
+      Collection<Artifact> missing = new ArrayList<>();
+      for (Artifact artifact : request.getManifest().getArtifactList()) {
+        // TODO: Validate the checksums on the server side, to fail more aggressively if require
+        if (!getArtifactFile(artifact.getName()).exists()) {
+          missing.add(artifact);
+        }
+      }
+      if (!missing.isEmpty()) {
+        throw Status.INVALID_ARGUMENT
+            .withDescription(
+                String.format("Attempted to commit manifest with missing Artifacts: [%s]", missing))
+            .asRuntimeException();
+      }
+      File mf = new File(stagingBase, "MANIFEST");
+      checkState(mf.createNewFile(), "Could not create file to store manifest");
+      try (OutputStream mfOut = new FileOutputStream(mf)) {
+        request.getManifest().writeTo(mfOut);
+      }
+      responseObserver.onNext(
+          CommitManifestResponse.newBuilder()
+              .setStagingToken(stagingBase.getCanonicalPath())
+              .build());
+      responseObserver.onCompleted();
+    } catch (StatusRuntimeException e) {
+      responseObserver.onError(e);
+      LOG.error("Failed to commit Manifest {}", request.getManifest(), e);
+    } catch (Exception e) {
+      responseObserver.onError(
+          Status.INTERNAL
+              .withCause(e)
+              .withDescription(Throwables.getStackTraceAsString(e))
+              .asRuntimeException());
+      LOG.error("Failed to commit Manifest {}", request.getManifest(), e);
+    }
+  }
+
+  File getArtifactFile(String artifactName) {
+    return new File(artifactsBase, artifactName);
+  }
+
+  private class CreateAndWriteFileObserver implements StreamObserver<PutArtifactRequest> {
+    private final StreamObserver<PutArtifactResponse> responseObserver;
+    private FileWritingObserver writer;
+
+    private CreateAndWriteFileObserver(StreamObserver<PutArtifactResponse> responseObserver) {
+      this.responseObserver = responseObserver;
+    }
+
+    @Override
+    public void onNext(PutArtifactRequest value) {
+      try {
+        if (writer == null) {
+          if (!value.getContentCase().equals(ContentCase.NAME)) {
```

Comment:
To make this read more easily, it may be worth building little helpers like `checkGrpcArgument` which functions just like `checkArgument` but throws the appropriate kind of exception. Or does the stub possibly convert them?

Reply:
I'll file a JIRA for this